crates/core/src/factories/core_bootstrap.rs

```diff
@@ -0,0 +1,263 @@
+            peer_store,
+        ));
+
+        Self {
+            space,
+            push_send,
+            push_task,
+            poll_task,
+        }
+    }
+}
+
+impl Bootstrap for CoreBootstrap {
+    fn put(&self, info: Arc<agent::AgentInfoSigned>) {
+        // ignore puts outside our space.
+        if info.space != self.space {
+            tracing::error!(
+                ?info,
+                "Logic Error: Attempting to put an agent outside of this space"
+            );
+            return;
+        }
+
+        // if we can't push onto our large buffer channel... we've got problems
+        if let Err(err) = self.push_send.try_send(info) {
+            tracing::warn!(?err, "Bootstrap overloaded, dropping put");
+        }
+    }
+}
+
+async fn push_task(
+    config: CoreBootstrapConfig,
+    server_url: Arc<str>,
+    push_send: PushSend,
+    mut push_recv: PushRecv,
+) {
+    let mut wait = None;
+
+    while let Some(info) = push_recv.recv().await {
+        let url =
+            format!("{server_url}/bootstrap/{}/{}", &info.space, &info.agent);
+        let enc = match info.encode() {
+            Err(err) => {
+                tracing::error!(?err, "Could not encode agent info, dropping");
+                continue;
+            }
+            Ok(enc) => enc,
+        };
+        match tokio::task::spawn_blocking(move || {
+            ureq::put(&url).send_string(&enc)
+        })
+        .await
+        {
+            Ok(Ok(_)) => {
+                // the put was successful, we don't need to wait
+                // before sending the next info if it is ready
+                wait = None;
+            }
+            _ => {
+                let now = Timestamp::now();
+
+                // the put failed, send it back to try again if not expired
+                if info.expires_at > now {
+                    let _ = push_send.try_send(info);
+                }
+
+                // we need to configure a backoff so we don't hammer the server
+                match wait {
+                    None => wait = Some(config.backoff_min()),
+                    Some(p) => {
+                        let mut p = p * 2;
+                        if p > config.backoff_max() {
+                            p = config.backoff_max();
+                        }
+                        wait = Some(p);
+                    }
+                }
+
+                // wait for the backoff time
+                if let Some(wait) = &wait {
+                    tokio::time::sleep(*wait).await;
+                }
+            }
+        }
+    }
+}
+
+async fn poll_task(
+    builder: Arc<builder::Builder>,
+    config: CoreBootstrapConfig,
+    server_url: Arc<str>,
+    space: SpaceId,
+    peer_store: peer_store::DynPeerStore,
+) {
+    let mut wait = config.backoff_min();
+
+    loop {
+        let url = format!("{server_url}/bootstrap/{space}");
+        if let Ok(Ok(data)) = tokio::task::spawn_blocking(move || {
```

Comment:
Should failure be logged here, in case the polling fails continuously and no one notices?

Reply:
Done: https://github.com/holochain/kitsune2/pull/50/commits/2ddd8fe0a2bf0798767993c30e306957b2520138